### ISSUE TITLE
[BUG]: Loss saved in hall of fame is not identical to loss recalculated based on prediction

### ISSUE THREAD

**Issue by BrotherHa** [MilesCranmer/PySR#816]
### What happened?

Hello,

I have identified an issue with some PySR regressions, that I ran. The loss saved in the hall of fame is not identical to the loss calculated with the same dataset and same custom loss function based on the PySR prediction. I have documented the issue in the notebook file attached.
Have similar problems occurred before or do you have any idea, what might be the reason for this?
As an additional test, i was thinking about trying to recalculate the loss of the individual inside PySR, the be able to reproduce the potentially faulty loss-calculation. Is there a way to do so?

Best regards and thanks in advance!

[loss_difference_pysr.pdf](https://github.com/user-attachments/files/18537891/loss_difference_pysr.pdf)

### Version

1.3.1

### Operating System

Windows

### Package Manager

Conda

### Interface

Other (specify below)

### Relevant log output

```shell

```

### Extra Info

[loss_difference_pysr.pdf](https://github.com/user-attachments/files/18537892/loss_difference_pysr.pdf)

**Comment by MilesCranmer** [comment]
@BrotherHa is it okay if you paste the code inline in the markdown block? This is so that search indexes like Google can correctly find this issue if other people have a similar problem.

**Comment by BrotherHa** [comment]
Hello @MilesCranmer 
Thank you very much for the quick reply. Is the format ok like that?
Best regards

**Comment by BrotherHa** [comment]
I have searched a bit where the issue occurs and in fact for most individuals of the regression the loss values from the equations_-table and the recalculated ones are absolutely identical until very high precision, but for some rare cases they divert significantly as in the example.

**Comment by MilesCranmer** [comment]
Is it only when reloading from a pickle file?

**Comment by MilesCranmer** [comment]
And are you sure you are computing the loss in a way completely identical between the Julia and the Python side of things?

**Comment by BrotherHa** [comment]
About the loss functions I am very sure. The only issues here might be problems caused by the precision of calculation, etc. And for more than 90% of all individuals both methods calculate exactly the same loss.
And concerning the pickle-file, I have also tried using the 'from_file' method, but it still stays the same.

**Comment by MilesCranmer** [comment]
> 90% of all individuals both methods calculate exactly the same loss

Do you see any pattern in the ones that don’t match? e.g., Are they very complex expressions?

**Comment by BrotherHa** [comment]
> > 90% of all individuals both methods calculate exactly the same loss
> 
> Do you see any pattern in the ones that don’t match? e.g., Are they very complex expressions?

I do not see any pattern in the equations, that have the error. They are not the most complex ones neither do they have explicitly complex operators.

**Comment by MilesCranmer** [comment]
But it's _only_ when you have a model loaded from disk, right? Like if you did:

```python
model = PySRRegressor(warm_start=True)  # fresh model; NOT loaded from disk

model.fit(X, y)

#= ... =#

model.fit(X, y)
```

**Comment by BrotherHa** [comment]
The Problem occured in the first place when fitting a new model. Then directly after the regression finished I called predict() and calculated the loss for training, test and total dataset based on that. But before that I saved the output of model.get_best(). The value from get_best() and the manual calculation for the best individual shows the bahavior in a less significant way (0.014545 instead of 0.014536), but it is still not in the range most other values are equal. For the example given in my first messege here I called the predict() method also just after the initial run and calculated the loss as 0.022473989882165576, but I did not save the loss from mode.equations_ separately. But in the hall of fame csv-file loss is also saved as 0.01720243, like it is outputed when I call model.equations_ from the loaded model.

**Comment by AllanChain** [comment]
I have encountered a similar problem when using a custom cosine-similarity-based loss.

<details>
<summary>Example Code</summary>

```python
import numpy as np
import pandas as pd
from pysr import PySRRegressor

nsamples = 30
np.random.seed(4283223)
x1 = np.random.uniform(size=nsamples)
x2 = np.random.uniform(size=nsamples)
x3 = np.random.uniform(size=nsamples)
x4 = np.random.uniform(size=nsamples)

X = pd.DataFrame({"x_1": x1, "x_2": x2, "x_3": x3, "x_4": x4})
target = (x1 * x4 - x3 * x2) ** 3

model = PySRRegressor(
    random_state=4208438902,
    deterministic=True,
    parallelism="serial",
    niterations=30,
    binary_operators=["+", "-", "*", "/"],
    unary_operators=["cube", "square"],
    loss_function="""
using Statistics

function eval_loss(tree, dataset::Dataset{T,L}, options)::L where {T,L}
    prediction, _ = eval_tree_array(tree, dataset.X, options)
    ratio = prediction ./ dataset.y
    norm = sqrt(sum(abs2, prediction) * sum(abs2, dataset.y))
    cos_value = sum(prediction .* dataset.y) / norm
    loss = abs(acos(clamp(cos_value, 0.0, 1.0)))
    scale = mean(abs, prediction)
    scale_loss = scale + 1 / (scale + 1e-6)
    return L(loss + 1e-6 * scale_loss)
end
""",
)

model.fit(X, target)
print(model)
print("Formula:", model.sympy())

prediction = model.predict(X)
norm = np.sqrt(np.sum(target**2) * np.sum(prediction**2))
cos_value = np.sum(target * prediction) / norm
loss = np.abs(np.arccos(np.clip(cos_value, 0, 1)))
scale = np.mean(np.abs(prediction))
scale_loss = scale + 1 / (scale + 1e-6)
print("Computed loss in Python:", loss + 1e-6 * scale_loss)
```
</details>

The target function in this example is `(x_1*x_4 - x_2*x_3)**3`, and I want to maximize the cosine similarity between the prediction and the target. To prevent the coefficient becomes too large and introduce numerical instabilities, I added the `scale_loss`. Set aside the validity of the loss design, in the above example, the reported loss can be very different with the loss recalculated in Python. For example:

```
PySRRegressor.equations_ = [
	  pick      score                    equation      loss  complexity
	0         0.000000                2.642676e-25  1.000000           1
	1        11.553640          cube(1.9243316e22)  0.000010           2
	2  >>>>   1.328369  square(cube(1.9243316e22))  0.000003           3
]
Formula: 5.07784534534534e+133
Computed loss in Python: 5.077845345345341e+127
```

The reported loss is very small, but it should be a very large one.

Although I have specified the random seed and make PySR deterministic, the final result is still random. 

<details>
<summary>Here are outputs of all five runs</summary>

```
PySRRegressor.equations_ = [
	  pick      score                    equation      loss  complexity
	0         0.000000                2.642676e-25  1.000000           1
	1        11.553640          cube(1.9243316e22)  0.000010           2
	2  >>>>   1.328369  square(cube(1.9243316e22))  0.000003           3
]
Formula: 5.07784534534534e+133
Computed loss in Python: 5.077845345345341e+127
```
```
PySRRegressor.equations_ = [
	  pick     score                                       equation      loss  complexity
	0        0.000000                                       7.33e-43  1.000000           1
	1        0.987759                                1.1066597 / 0.0  0.138689           3
	2        1.754098                cube((x_1 * x_4) - (x_2 * x_3))  0.000022           8
	3  >>>>  1.188177  cube(((x_1 * x_4) - (x_2 * x_3)) * 2.7796757)  0.000002          10
]
Formula: 21.4774339167278*(x_1*x_4 - x_2*x_3)**3
Computed loss in Python: 1.9999991259531853e-06
```
```
PySRRegressor.equations_ = [
	  pick      score                                           equation      loss  complexity
	0         0.000000                                      1.3934125e-37  1.000000           1
	1        11.240707                                 cube(1.3326811e38)  0.000013           2
	2         0.557888                         cube(square(1.3326811e38))  0.000008           3
	3  >>>>   0.611431             cube(square(1.3326811e38)) - 0.9821605  0.000002           5
	4         0.008833  (cube(square(1.3326811e38)) - square(1.3070838...  0.000002           8
]
Formula: 5.60218481169532e+228 - 1*0.9821605
/home/ac/Projects/PKU/HallSR/pysr_bug.py:43: RuntimeWarning: overflow encountered in square
  norm = np.sqrt(np.sum(target**2) * np.sum(prediction**2))
Computed loss in Python: 5.6021848116953185e+222
```
```
PySRRegressor.equations_ = [
	  pick     score                                           equation      loss  complexity
	0        0.000000                                      1.5459748e-30  1.000000           1
	1        0.057146                                    cube(cube(x_4))  0.891998           3
	2        0.302634                                    cube(x_1 * x_4)  0.659070           4
	3        0.091833                              cube(cube(x_1 + x_4))  0.601242           5
	4        0.398972                     cube(x_4 * (x_1 - 0.44725403))  0.403439           6
	5        0.331675                      cube(x_4 * (x_1 - cube(x_3)))  0.289556           7
	6        9.506604                    cube((x_4 * x_1) - (x_3 * x_2))  0.000022           8
	7  >>>>  1.188146      cube(((x_1 * x_4) - (x_3 * x_2)) / 0.3583628)  0.000002          10
	8        0.000006  (cube((x_4 * x_1) - (x_3 * x_2)) / 0.010429359...  0.000002          12
	9        0.000026  ((cube((x_1 * x_4) - (x_2 * x_3)) * 0.83377784...  0.000002          14
]
Formula: 21.7285745471395*(x_1*x_4 - x_2*x_3)**3
Computed loss in Python: 2.000126025169295e-06
```
```
PySRRegressor.equations_ = [
	  pick      score                                           equation      loss  complexity
	0         0.000000                                      1.0714004e-24  1.000000           1
	1         0.780003                                 cube(5.0544116e29)  0.458405           2
	2        10.262700                           cube(cube(5.0544116e29))  0.000016           3
	3         0.082098                     cube(cube(cube(5.0544116e29)))  0.000015           4
	4  >>>>   0.375597  square(cube(square(square(cube(cube(1.3362112e...  0.000002           9
	5         0.016263  square(cube(3.4662442e29) / cube(x_4 / cube(-1...  0.000002          16
]
Formula: 1.54533717529716e+8235*x_3
/home/ac/Projects/PKU/HallSR/.venv/lib/python3.12/site-packages/numpy/_core/fromnumeric.py:86: RuntimeWarning: invalid value encountered in reduce
  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)
Computed loss in Python: nan
```
</details>

**Comment by MilesCranmer** [comment]
@AllanChain could you set `precision=64` and try again? The backend defaults to 32-bit for speed

**Comment by MilesCranmer** [comment]
I suspect it’s related to not handling the second argument of `eval_tree_loss` in your custom loss function. The second argument tells you if the evaluation succeeded or not - if not, then the first argument is basically random, and should not be used. This is used to deal with stuff like Inf and NaN in a safe way

**Comment by AllanChain** [comment]
Thank you! It's my bad not looking into the documentation carefully, and I'm sorry for bringing in an unrelated issue.
